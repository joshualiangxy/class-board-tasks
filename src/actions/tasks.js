import firebase from '../firebase/firebase';

export const addPersonalTask = task => ({ type: 'ADD_PERSONAL_TASK', task });

export const startAddPersonalTask = task => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const id = task.id;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .set(task)
      .then(() => dispatch(addPersonalTask(task)));
  };
};

export const removePersonalTask = id => ({ type: 'REMOVE_PERSONAL_TASK', id });

export const startRemovePersonalTask = id => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .delete()
      .then(() => dispatch(removePersonalTask(id)));
  };
};

export const editPersonalTask = (id, updates) => ({
  type: 'EDIT_PERSONAL_TASK',
  id,
  updates
});

export const startEditPersonalTask = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .set(updates)
      .then(dispatch(editPersonalTask(id, updates)));
  };
};

export const setTasks = tasks => ({ type: 'SET_TASK', tasks });

export const startSetTasks = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const tasks = [];
    const userRef = firebase.firestore().collection('users').doc(uid);
    const tasksRef = userRef.collection('tasks');

    return tasksRef
      .get()
      .then(querySnapshot =>
        querySnapshot.forEach(queryDocSnapshot =>
          tasks.push({ id: queryDocSnapshot.id, ...queryDocSnapshot.data() })
        )
      )
      .then(() => {
        userRef.get().then(userDocSnapshot => {
          const groupIds = userDocSnapshot.get('groups') || [];
          const groupsPromises = [];

          groupIds.forEach(gid => {
            const groupPromises = [];
            const groupRef = firebase.firestore().collection('groups').doc(gid);

            groupPromises.push(
              groupRef
                .get()
                .then(groupDocSnapshot => groupDocSnapshot.get('name'))
            );
            groupPromises.push(groupRef.collection('tasks').get());

            groupsPromises.push(
              Promise.all(groupPromises).then(
                ([groupName, queryTaskCollectionSnapshot]) =>
                  queryTaskCollectionSnapshot.forEach(queryTaskDocSnapshot =>
                    tasks.push({
                      groupName,
                      gid,
                      id: queryTaskDocSnapshot.id,
                      ...queryTaskDocSnapshot.data()
                    })
                  )
              )
            );

            return Promise.all(groupPromises);
          });
        });
      })
      .then(() => dispatch(setTasks(tasks)));
  };
};

export const toggleCompletedPersonal = (id, completedState) => ({
  type: 'TOGGLE_COMPLETED_PERSONAL',
  id,
  completedState
});

export const startToggleCompletedPersonal = (id, completedState) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .update({ completed: !completedState })
      .then(() => toggleCompletedPersonal(id, completedState));
  };
};

export const removeTaskData = () => ({ type: 'REMOVE_TASK_DATA' });
