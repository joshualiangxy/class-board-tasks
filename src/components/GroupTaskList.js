import React, { useState, useEffect } from 'react';
import TaskListItem from './TaskListItem';
import getSortedTasks from '../selectors/tasks';
import TaskListHeader from './TaskListHeader';

const GroupTaskList = ({
  removeUserFromTask,
  removeAdmin,
  tasks,
  users,
  admin,
  editGroupTask,
  removeGroupTask,
  toggleGroupTaskComplete,
  groupName
}) => {
  const [sortedList, setSortedList] = useState(getSortedTasks(tasks));

  useEffect(() => {
    setSortedList(getSortedTasks(tasks));
  }, [tasks]);

  return (
    <div className="content-container list">
      <TaskListHeader isGroup={true} />
      {sortedList.others.tasks.length === 0 ? (
        <div className="list-item__header list-item__header-message">
          <p>No tasks yet</p>
        </div>
      ) : (
        Object.values(sortedList)[0].tasks.map(task => (
          <TaskListItem
            key={task.id}
            removeUserFromTask={removeUserFromTask}
            removeAdmin={removeAdmin}
            users={users}
            task={task}
            showGroup={false}
            admin={admin}
            groupName={groupName}
            editGroupTask={editGroupTask}
            removeGroupTask={removeGroupTask}
            toggleGroupTaskComplete={toggleGroupTaskComplete}
          />
        ))
      )}
    </div>
  );
};

export default GroupTaskList;
