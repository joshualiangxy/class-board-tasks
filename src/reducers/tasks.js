const tasksReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_TASK':
      return [...state, action.task];
    case 'REMOVE_TASK':
      return state.filter(task => task.id !== action.id);
    case 'EDIT_TASK':
      return state.map(task =>
        task.id === action.id
          ? {
              ...task,
              ...action.updates
            }
          : task
      );
    case 'SET_TASK':
      return action.tasks;
    default:
      return state;
  }
};

export default tasksReducer;
