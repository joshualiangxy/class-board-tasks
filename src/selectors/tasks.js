const getSortedTasks = (tasks, { text, sortBy, grouped }) => {
  const lowercaseText = text.toLowerCase();

  const processedTasks = tasks
    .filter(
      task =>
        task.title.toLowerCase().includes(lowercaseText) ||
        task.description.toLowerCase().includes(lowercaseText) ||
        task.module.toLowerCase().includes(lowercaseText)
    )
    .sort((taskOne, taskTwo) => {
      switch (sortBy) {
        case 'deadline':
          if (taskOne.deadline && taskTwo.deadline)
            return taskOne.deadline - taskTwo.deadline;
          else if (taskOne.deadline) return -1;
          else if (taskTwo.deadline) return 1;
          else return taskOne.title.localeCompare(taskTwo.title);
        case 'name':
          return taskOne.title.localeCompare(taskTwo.title);
      }
    });

  const sortedList = {};

  if (grouped) {
    processedTasks.forEach(task => {
      if (task.groupName) {
        const gid = task.gid;

        if (!sortedList[gid])
          sortedList[gid] = { groupName: task.groupName, tasks: [] };
        sortedList[gid].tasks.push(task);
      } else if (task.module) {
        const module = task.module;

        if (!sortedList[module])
          sortedList[module] = { groupName: task.module, tasks: [] };
        sortedList[module].tasks.push(task);
      } else {
        if (!sortedList.others)
          sortedList.others = { groupName: 'others', tasks: [] };
        sortedList.others.tasks.push(task);
      }
    });
  } else sortedList.others = { groupName: 'others', tasks: processedTasks };

  return sortedList;
};

export default getSortedTasks;
