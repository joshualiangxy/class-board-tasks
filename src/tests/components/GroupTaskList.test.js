import React from 'react';
import { shallow } from 'enzyme';
import GroupTaskList from '../../components/GroupTaskList';
import { singleGroup as tasks } from '../fixtures/tasks';
import users from '../fixtures/groupUsers';

const wrapper = shallow(<GroupTaskList tasks={tasks} users={users} />);

describe('render', () => {
  it('should render GroupTaskList', () => expect(wrapper).toMatchSnapshot());

  it('should render GroupTaskList without tasks', () => {
    wrapper.setProps({ tasks: [] });

    expect(wrapper).toMatchSnapshot();
  });
});
