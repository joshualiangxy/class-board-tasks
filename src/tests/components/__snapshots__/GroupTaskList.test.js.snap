// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render GroupTaskList 1`] = `
<div
  className="content-container list"
>
  <Connect(TaskListHeader)
    isGroup={true}
  />
  <Connect(TaskListItem)
    key="c1511c73-07b9-44cd-a4bd-01758515acb5"
    showGroup={false}
    task={
      Object {
        "completed": Object {
          "testuid": false,
        },
        "deadline": 1593489600000,
        "description": "Finish readings",
        "downloadURLs": Object {},
        "gid": "72b6b3bb-cb74-4d08-ae67-dfc1b51baaf9",
        "id": "c1511c73-07b9-44cd-a4bd-01758515acb5",
        "module": "CS1231",
        "title": "Tutorial",
        "uploadRequired": true,
      }
    }
    users={
      Array [
        Object {
          "admin": true,
          "displayName": "Robert",
          "studentNum": "A0123456A",
          "uid": "testuid",
        },
        Object {
          "admin": false,
          "displayName": "Peter",
          "studentNum": "A0000000A",
          "uid": "testuid2",
        },
      ]
    }
  />
  <Connect(TaskListItem)
    key="076e610e-aade-4d2b-897f-81a29f00c76b"
    showGroup={false}
    task={
      Object {
        "completed": Object {
          "testuid": false,
        },
        "description": "Finish by Sunday!",
        "downloadURLs": Object {
          "testuid": Object {
            "downloadURL": "website.com",
            "fileName": "Homework.pdf",
          },
        },
        "gid": "72b6b3bb-cb74-4d08-ae67-dfc1b51baaf9",
        "id": "076e610e-aade-4d2b-897f-81a29f00c76b",
        "module": "CS1231",
        "title": "Homework 1",
        "uploadRequired": true,
      }
    }
    users={
      Array [
        Object {
          "admin": true,
          "displayName": "Robert",
          "studentNum": "A0123456A",
          "uid": "testuid",
        },
        Object {
          "admin": false,
          "displayName": "Peter",
          "studentNum": "A0000000A",
          "uid": "testuid2",
        },
      ]
    }
  />
  <Connect(TaskListItem)
    key="0b80e917-78a4-4f87-a379-cdc1733923d6"
    showGroup={false}
    task={
      Object {
        "completed": Object {},
        "deadline": 1593403200000,
        "description": "",
        "gid": "72b6b3bb-cb74-4d08-ae67-dfc1b51baaf9",
        "id": "0b80e917-78a4-4f87-a379-cdc1733923d6",
        "module": "CS1231",
        "title": "Watch lectures",
      }
    }
    users={
      Array [
        Object {
          "admin": true,
          "displayName": "Robert",
          "studentNum": "A0123456A",
          "uid": "testuid",
        },
        Object {
          "admin": false,
          "displayName": "Peter",
          "studentNum": "A0000000A",
          "uid": "testuid2",
        },
      ]
    }
  />
</div>
`;

exports[`render should render GroupTaskList without tasks 1`] = `
<div
  className="content-container list"
>
  <Connect(TaskListHeader)
    isGroup={true}
  />
  <div
    className="list-item__header list-item__header-message"
  >
    <p>
      No tasks yet
    </p>
  </div>
</div>
`;
